internal/tracker/upstream.go

```diff
@@ -0,0 +1,196 @@
+package tracker
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+var errorNoAvailableUpstream = errors.New("No Available Upstream")
+
+// UpstreamConns tracks connections for an upstreamGroup
+// on a per upstream basis. Upstreams can be marked as
+// unhealthy to prevent them from being chosen for new connections.
+// UpstreamConns handles load balancing through BeginConnection()
+type UpstreamConns struct {
+	// mu protects the resources of UpstreamConns
+	mu sync.Mutex
+
+	// upstreams holds all upstreams, healthy or unhealthy
+	upstreams map[uuid.UUID]*upstream
+
+	// pq holds healthy upstreams and provides the means to
+	// pick the upstream with the least connections.
+	pq *upstreamPQ
+}
+
+// An upstream stores a count of connections
+// as well as some overhead for use in upstreamPQ.
+type upstream struct {
+	// id is the id of the upstream
+	id uuid.UUID
+
+	// The count of connections to the upstream.
+	// Also the priority of an upstream, lowest first.
+	connCount uint32
+
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	// if an upstream is pulled from the upstreamPQ (because of health)
+	// its index will be set to -1
+	index int
+}
+
+// NewUpstreamConns creates a new UpstreamConns
+// with upstreams based on provided upstreamIDs.
+// upstreams must be marked as healthy before they will be
+// added to the internal priorityQueue and available for BeginConnection()
+func NewUpstreamConns(upstreamIDs []uuid.UUID) *UpstreamConns {
+	upstreams := map[uuid.UUID]*upstream{}
```

Comment:
Done. Out of curiosity, do you prioritize changes like this as a principle so you cover critical cases by default, or do you feel this case it is critical?

Reply:
Yea, sounds like we are on the same page. Since this line is executed at startup time, I don't think its impact on performance is all that significant, but I think its always good to be in the habit of allocating based on expected or known capacity. I took the suggestion 😄